server/common/diffexpdu/postingslist.py

```diff
@@ -0,0 +1,426 @@
+    """
+    Encode lower two bytes only. Applies steps:
+        * optional inversion
+        * delta codes
+        * byte shuffle
+        * zlib.compress
+    """
+    key = part.arr[part.start_idx] >> 16
+    n_elem = len(arr)
+
+    if _interval_inverted_length(arr) + 2 >= n_elem:
+        block_type = _BlockDescription.BlockType.Uint16Array
+        arr = _byteshuffle(_delta(arr))
+        buf = arr.tobytes()
+        assert len(buf) == 2 * n_elem
+    else:
+        # invert the array if it would be shorter to encode the
+        # range over which it is inverted and the elements not
+        # in the original
+        block_type = _BlockDescription.BlockType.Uint16ArrayInverted
+        buf = bytearray(4)
+        uint16.pack_into(buf, 0, int(arr[0]))  # range start
+        uint16.pack_into(buf, 2, int(arr[-1]))  # range end INCLUSIVE
+        arr = _byteshuffle(_delta(_interval_invert(arr)))
+        buf.extend(arr.tobytes())
+        assert len(buf) == 4 + len(arr)
+
+    buf = _blockCompress(buf)
+    header = _BlockDescription(
+        block_type=block_type, list_id_mask=(1 << part.list_id), n_elem=n_elem, key=key, n_bytes=len(buf)
+    )
+    return (header, buf)
+
+
+def inflate_postings_lists(buf: Union[bytes, bytearray, memoryview], offset=0) -> List[np.ndarray]:
+    """
+    Decode a serialized postings list, returning the original list as a ``numpy.ndarray``.
+    """
+    buf = memoryview(buf)
+    n_lists, block_descriptions = _decode_header(buf[offset:])
+    n_blocks = len(block_descriptions)
+    offset += headerPacker.size + n_blocks * _BlockDescription.packed_length
+
+    # decode each block
+    sub_lists = []
+    for block_description in block_descriptions:
+        sub_lists.extend(_decode_block(block_description, buf[offset : offset + block_description.n_bytes]))
+        offset += block_description.n_bytes
+
+    sub_lists.sort(key=lambda sl: sl[0])  # sort by list id
+    concat_lists = [
+        np.concatenate([g[1] for g in grp]) for _, grp in itertools.groupby(sub_lists, key=lambda sl: sl[0])
+    ]
+    return concat_lists
+
+
+def _decode_header(buf: memoryview) -> Tuple[int, List[_BlockDescription]]:
+    magic, n_lists, n_blocks = headerPacker.unpack_from(buf)
+    offset = headerPacker.size
+    n_blocks += 1
+    n_lists += 1
+    assert magic == MAGIC_NUMBER  # TODO: convert most asserts to exceptions
+    assert n_blocks > 0
+    assert n_lists > 0
+
+    # decode block descriptions
+    block_descriptions = [
+        _BlockDescription.from_bytes(buf[b:])
+        for b in range(
+            offset,
+            offset + n_blocks * _BlockDescription.packed_length,
+            _BlockDescription.packed_length,
+        )
+    ]
+    return n_lists, block_descriptions
+
+
+def _decode_block(desc: _BlockDescription, buf: memoryview) -> List[Tuple[ListId, np.ndarray]]:
+    block_type = desc.block_type
+    n_elem = desc.n_elem
+    buf = _blockDecompress(buf)
+    list_ids = desc.list_ids()
```

Comment:
I thought each block contains values from a single list? Indeed when decoding the blocks you have `assert len(list_ids) == 1`

Reply:
True, for now.  This extra feature is in place to allow multi-list blocks if we find they have value.  I prototyped it, and did not find a big win, so I left out the actual block type (but retained the framework).

Currently, any single BitArray or Uint16 block can contain a single list - the list is identified by the "list ID mask" which can contain only a single bit set.   A diffexp PDU will have blocks from multiple lists, with their specific list ID encoded in this mask (currently diffex uses 2 lists, but the format supports up to 8 as we may find other uses for it).

We could simplify this slightly by changing the list "mask" to a list "id", but there is no major value in doing so, and it would remove the ability to add multi-list blocks in the future.

Reply:
I added a note about this to the docs as well.